bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -900,23 +898,30 @@ public void publish(String topic, byte[] payload, MqttActionCallback listener) {
      * @param payload The message payload
      * @param qos The quality of service for this message
      * @param retain Set to true to retain the message on the broker
-     * @param listener An optional listener to be notified of success or failure of the delivery.
      * @return Returns a future that completes with a result of true if the publishing succeeded and completes
      *         exceptionally on an error or with a result of false if no broker connection is established.
      */
-    public CompletableFuture<Boolean> publish(String topic, byte[] payload, int qos, boolean retain) {
-        MqttAsyncClient client = this.client;
+    public CompletableFuture<Boolean> publish(String topic, byte[] payload, @Nullable Integer qos, @Nullable Boolean retain) {
+        final MqttAsyncClientWrapper client = this.client;
         if (client == null) {
             return CompletableFuture.completedFuture(false);
         }
+
+        // use defaults if not provided
+        qos = qos == null ? this.qos : qos;
+        retain = retain == null ? this.retain : retain;
+
         // publish message asynchronously
-        CompletableFuture<Boolean> f = new CompletableFuture<>();
-        try {
-            client.publish(topic, payload, qos, retain, f, actionCallback);
-        } catch (org.eclipse.paho.client.mqttv3.MqttException e) {
-            f.completeExceptionally(new MqttException(e));
-        }
-        return f;
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        client.publish(topic, payload, retain, qos).whenComplete((m, t) -> {
```

Comment:
Why can't you use the clients future directly as return value? Maybe add the (unnecessary) boolean argument to the future to make it compatible?

This conversion stuff happening here right now is probably more inefficient than carrying a useless boolean in the type signature IMO.

Reply:
The clients return different futures: CompletableFuture<Mqtt3Publish> and CompletableFuture<Mqtt5Publish>.